File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.empty()){
            return 0;
        }
        int longest=0;
        unordered_map<int,int> occurences;
        int left=0;
        for(int right=0;right<s.size();++right){
            char currChar=s[right];
            if(occurences.find(currChar)!=occurences.end()&&occurences[currChar]>=left){
                left=occurences[currChar]+1;
            }
            occurences[currChar]=right;
            longest=max(longest,right-left+1);
        }
        return longest;
    }
};

// class Solution {
// public:
//     int lengthOfLongestSubstring(string s) {
//         int n=s.size();
//         unordered_set<int> seen;
//         int maxLength=0;
//         int start=0;

//         for(int end=0;end<n;++end){
//             while(seen.find(s[end])!=seen.end()){
//                 seen.erase(s[start]);
//                 start++;
//             }
//             seen.insert(s[end]);
//             maxLength=max(maxLength,end-start+1);
//         }
//         return maxLength;
        
//     }
// };

File: 3. Longest Substring Without Repeating Characters.rs
use std::collections::HashMap;
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut map = HashMap::new();
        let mut left = 0;
        let mut ans = 0;

        for right in 0..chars.len() {
            if let Some(&index) = map.get(&chars[right]) {
                left = left.max(index + 1);
            }
            map.insert(chars[right], right);
            ans = ans.max(right - left + 1);
        }

        ans as i32
    }
}
